core/lib/__tests__/matchers.ts

```diff
@@ -0,0 +1,76 @@
+import { expect } from "vitest";
+
+import type { ProcessedPub } from "contracts";
+import type { PubsId } from "db/public";
+
+import type { db } from "~/kysely/database";
+
+const deepSortValues = (pub: ProcessedPub): ProcessedPub => {
+	pub.values
+		.sort((a, b) => (a.value as string).localeCompare(b.value as string))
+		.map((item) => ({
+			...item,
+			relatedPub: item.relatedPub?.values ? deepSortValues(item.relatedPub) : item.relatedPub,
+		}));
+
+	return pub;
+};
+
+expect.extend({
+	async toExist(received: PubsId | ProcessedPub, expected?: typeof db) {
+		if (typeof received !== "string") {
+			throw new Error("toExist() can only be called with a PubsId");
+		}
+		const { getPlainPub } = await import("../server/pub");
+
+		const pub = await getPlainPub(received, expected).executeTakeFirst();
+		const pass = Boolean(pub && pub.id === received);
+		const { isNot } = this;
+
+		return {
+			pass,
+			message: () =>
+				isNot
+					? `Expected pub with ID ${received} not to exist, but it ${pass ? "does" : "does not"}`
+					: `Expected pub with ID ${received} to exist, but it ${pass ? "does" : "does not"}`,
+		};
+	},
+
+	toHaveValues(
+		received: PubsId | ProcessedPub,
+		expected: Partial<ProcessedPub["values"][number]>[]
+	) {
+		if (typeof received === "string") {
+			throw new Error("toHaveValues() can only be called with a ProcessedPub");
```

Comment:
how come `received: PubsId | ProcessedPub` up above instead of `received: ProcessedPub` then?

Reply:
good catch! it was left over bc i think i wanted to be able to pass in both in either case, and was having some trouble with the types. i later decided to just use `PubsId` for toExist and `ProcessedPub` for toHaveValues, but never updated it! I'll fix it